File: trafilatura/sitemaps.py
"""
Deriving link info from sitemaps.
"""

import logging
import re

from itertools import islice
from time import sleep
from typing import List, Set, Optional

from courlan import (
    clean_url,
    extract_domain,
    filter_urls,
    fix_relative_urls,
    get_hostinfo,
    lang_filter,
)

from .downloads import fetch_url, is_live_page
from .settings import MAX_LINKS, MAX_SITEMAPS_SEEN
from .utils import is_similar_domain

# import urllib.robotparser # Python >= 3.8
# ROBOT_PARSER = urllib.robotparser.RobotFileParser()


LOGGER = logging.getLogger(__name__)

LINK_REGEX = re.compile(r"<loc>(?:<!\[CDATA\[)?(http.+?)(?:\]\]>)?</loc>")
XHTML_REGEX = re.compile(r"<xhtml:link.+?>", re.DOTALL)
HREFLANG_REGEX = re.compile(r'href=["\'](.+?)["\']')
WHITELISTED_PLATFORMS = re.compile(
    r"(?:blogger|blogpost|ghost|hubspot|livejournal|medium|typepad|squarespace|tumblr|weebly|wix|wordpress)\."
)

SITEMAP_FORMAT = re.compile(r"^.{0,5}<\?xml|<sitemap|<urlset")
DETECT_SITEMAP_LINK = re.compile(r"\.xml(\..{2,4})?$|\.xml[?#]")
DETECT_LINKS = re.compile(r'https?://[^\s<"]+')
SCRUB_REGEX = re.compile(r"\?.*$|#.*$")
POTENTIAL_SITEMAP = re.compile(r"\.xml\b")  # |\bsitemap\b

GUESSES = [
    "sitemap.xml",
    "sitemap.xml.gz",
    "sitemap",
    "sitemap_index.xml",
    "sitemap_news.xml",
]


class SitemapObject:
    "Store all necessary information on sitemap download and processing."
    __slots__ = [
        "base_url",
        "content",
        "current_url",
        "domain",
        "external",
        "seen",
        "sitemap_urls",
        "target_lang",
        "urls",
    ]

    def __init__(
        self,
        base_url: str,
        domain: str,
        sitemapsurls: List[str],
        target_lang: Optional[str] = None,
        external: bool = False,
    ) -> None:
        self.base_url: str = base_url
        self.content: str = ""
        self.domain: str = domain
        self.external: bool = external
        self.current_url: str = ""
        self.seen: Set[str] = set()
        self.sitemap_urls: List[str] = sitemapsurls
        self.target_lang: Optional[str] = target_lang
        self.urls: List[str] = []

    def fetch(self) -> None:
        "Fetch a sitemap over the network."
        LOGGER.debug("fetching sitemap: %s", self.current_url)
        self.content = fetch_url(self.current_url)
        self.seen.add(self.current_url)

    def handle_link(self, link: str) -> None:
        """Examine a link and determine if it's valid and if it leads to
        a sitemap or a web page."""
        if link == self.current_url:  # safety check
            return
        # fix, check, clean and normalize
        link = fix_relative_urls(self.base_url, link)
        link = clean_url(link, self.target_lang)

        if link is None or not lang_filter(link, self.target_lang):
            return

        newdomain = extract_domain(link, fast=True)
        if newdomain is None:
            LOGGER.error("couldn't extract domain: %s", link)
            return

        # don't take links from another domain and make an exception for main platforms
        # also bypass: subdomains vs. domains
        if (
            not self.external
            and not WHITELISTED_PLATFORMS.search(newdomain)
            and not is_similar_domain(self.domain, newdomain)
        ):
            LOGGER.warning(
                "link discarded, diverging domain names: %s %s", self.domain, newdomain
            )
            return

        if DETECT_SITEMAP_LINK.search(link):
            self.sitemap_urls.append(link)
        else:
            self.urls.append(link)

    def extract_sitemap_langlinks(self) -> None:
        "Extract links corresponding to a given target language."
        if "hreflang=" not in self.content:
            return
        # compile regex here for modularity and efficiency
        lang_regex = re.compile(
            rf"hreflang=[\"']({self.target_lang}.*?|x-default)[\"']", re.DOTALL
        )
        # extract
        for attrs in (
            m[0] for m in islice(XHTML_REGEX.finditer(self.content), MAX_LINKS)
        ):
            if lang_regex.search(attrs):
                lang_match = HREFLANG_REGEX.search(attrs)
                if lang_match:
                    self.handle_link(lang_match[1])
        LOGGER.debug(
            "%s sitemaps and %s links with hreflang found for %s",
            len(self.sitemap_urls),
            len(self.urls),
            self.current_url,
        )

    def extract_sitemap_links(self) -> None:
        "Extract sitemap links and web page links from a sitemap file."
        # extract
        for match in (
            m[1] for m in islice(LINK_REGEX.finditer(self.content), MAX_LINKS)
        ):
            # process middle part of the match tuple
            self.handle_link(match)
        LOGGER.debug(
            "%s sitemaps and %s links found for %s",
            len(self.sitemap_urls),
            len(self.urls),
            self.current_url,
        )

    def process(self) -> None:
        "Download a sitemap and extract the links it contains."
        plausible = is_plausible_sitemap(self.current_url, self.content)
        # safeguard
        if not plausible:
            return
        # try to extract links from TXT file
        if not SITEMAP_FORMAT.match(self.content):
            for match in (
                m[0] for m in islice(DETECT_LINKS.finditer(self.content), MAX_LINKS)
            ):
                self.handle_link(match)
            return
        # process XML sitemap
        if self.target_lang is not None:
            self.extract_sitemap_langlinks()
            if self.sitemap_urls or self.urls:
                return
        self.extract_sitemap_links()


def sitemap_search(
    url: str, target_lang: Optional[str] = None, external: bool = False, sleep_time: int = 2
) -> List[str]:
    """Look for sitemaps for the given URL and gather links.

    Args:
        url: Webpage or sitemap URL as string.
             Triggers URL-based filter if the webpage isn't a homepage.
        target_lang: Define a language to filter URLs based on heuristics
                     (two-letter string, ISO 639-1 format).
        external: Similar hosts only or external URLs
                  (boolean, defaults to False).
        sleep_time: Wait between requests on the same website.

    Returns:
        The extracted links as a list (sorted list of unique links).

    """
    domainname, baseurl = get_hostinfo(url)
    if domainname is None:
        LOGGER.warning("invalid URL: %s", url)
        return []

    if not is_live_page(baseurl):
        LOGGER.warning("base URL unreachable, dropping sitemap: %s", url)
        return []

    urlfilter = None

    if url.endswith((".gz", "sitemap", ".xml")):
        sitemapurls = [url]
    else:
        sitemapurls = []
        # set url filter to target subpages
        if len(url) > len(baseurl) + 2:
            urlfilter = url

    sitemap = SitemapObject(baseurl, domainname, sitemapurls, target_lang, external)

    # try sitemaps in robots.txt file, additional URLs just in case
    if not sitemap.sitemap_urls:
        sitemap.sitemap_urls = find_robots_sitemaps(baseurl) or [
            f"{baseurl}/{g}" for g in GUESSES
        ]

    # iterate through nested sitemaps and results
    while sitemap.sitemap_urls and len(sitemap.seen) < MAX_SITEMAPS_SEEN:
        sitemap.current_url = sitemap.sitemap_urls.pop()
        sitemap.fetch()
        sitemap.process()
        # sanity check: keep track of visited sitemaps and exclude them
        sitemap.sitemap_urls = [
            s for s in sitemap.sitemap_urls if s not in sitemap.seen
        ]

        if len(sitemap.seen) < MAX_SITEMAPS_SEEN:
            sleep(sleep_time)

    if urlfilter:
        sitemap.urls = filter_urls(sitemap.urls, urlfilter)

    LOGGER.debug("%s sitemap links found for %s", len(sitemap.urls), domainname)
    return sitemap.urls


def is_plausible_sitemap(url: str, contents: Optional[str]) -> bool:
    """Check if the sitemap corresponds to an expected format,
    i.e. TXT or XML."""
    if contents is None:
        return False

    # strip query and fragments
    url = SCRUB_REGEX.sub("", url)

    # check content
    if (
        POTENTIAL_SITEMAP.search(url)
        and (not isinstance(contents, str) or not SITEMAP_FORMAT.match(contents))
        or "<html" in contents[:150].lower()
    ):
        LOGGER.warning("not a valid XML sitemap: %s", url)
        return False

    return True


def find_robots_sitemaps(baseurl: str) -> List[str]:
    """Guess the location of the robots.txt file and try to extract
    sitemap URLs from it"""
    robotstxt = fetch_url(baseurl + "/robots.txt")
    return extract_robots_sitemaps(robotstxt, baseurl)


def extract_robots_sitemaps(robotstxt: str, baseurl: str) -> List[str]:
    "Read a robots.txt file and find sitemap links."
    # sanity check on length (cause: redirections)
    if robotstxt is None or len(robotstxt) > 10000:
        return []
    sitemapurls = []
    # source: https://github.com/python/cpython/blob/3.8/Lib/urllib/robotparser.py
    for line in robotstxt.splitlines():
        # remove optional comment and strip line
        i = line.find("#")
        if i >= 0:
            line = line[:i]
        line = line.strip()
        if not line:
            continue
        line = line.split(":", 1)
        if len(line) == 2:
            line[0] = line[0].strip().lower()
            if line[0] == "sitemap":
                # urllib.parse.unquote(line[1].strip())
                candidate = fix_relative_urls(baseurl, line[1].strip())
                sitemapurls.append(candidate)
    LOGGER.debug("%s sitemaps found in robots.txt", len(sitemapurls))
    return sitemapurls

File: trafilatura.js/sitemaps.js
// Deriving link info from sitemaps.

var { URL } = require('url');
var { cleanUrl, extractDomain, filterUrls, fixRelativeUrls, getHostinfo, langFilter } = require('./courlan');
var { isSimilarDomain } = require('./deduplication');
var { fetchUrl, isLivePage } = require('./downloads');
var { MAX_LINKS, MAX_SITEMAPS_SEEN } = require('./settings');

var LINK_REGEX = /<loc>(?:<!\[CDATA\[)?(http.+?)(?:\]\]>)?<\/loc>/g;
var XHTML_REGEX = /<xhtml:link.+?>/gs;
var HREFLANG_REGEX = /href=['"](.+?)['"]/;
var WHITELISTED_PLATFORMS = /(?:blogger|blogpost|ghost|hubspot|livejournal|medium|typepad|squarespace|tumblr|weebly|wix|wordpress)\./;

var SITEMAP_FORMAT = /^.{0,5}<\?xml|<sitemap|<urlset/;
var DETECT_SITEMAP_LINK = /\.xml(\..{2,4})?$|\.xml[?#]/;
var DETECT_LINKS = /https?:\/\/[^\s<"]+/g;
var SCRUB_REGEX = /\?.*$|#.*$/;
var POTENTIAL_SITEMAP = /\.xml\b/;

var GUESSES = [
    "sitemap.xml",
    "sitemap.xml.gz",
    "sitemap",
    "sitemap_index.xml",
    "sitemap_news.xml",
];

class SitemapObject {
    constructor(baseUrl, domain, sitemapUrls, targetLang = null, external = false) {
        this.baseUrl = baseUrl;
        this.content = "";
        this.domain = domain;
        this.external = external;
        this.currentUrl = "";
        this.seen = new Set();
        this.sitemapUrls = sitemapUrls;
        this.targetLang = targetLang;
        this.urls = [];
    }

    async fetch() {
        console.debug("fetching sitemap:", this.currentUrl);
        this.content = await fetchUrl(this.currentUrl);
        this.seen.add(this.currentUrl);
    }

    handleLink(link) {
        if (link === this.currentUrl) return;

        link = fixRelativeUrls(this.baseUrl, link);
        link = cleanUrl(link, this.targetLang);

        if (!link || !langFilter(link, this.targetLang)) return;

        var newDomain = extractDomain(link, { fast: true });
        if (!newDomain) {
            console.error("couldn't extract domain:", link);
            return;
        }

        if (!this.external && !WHITELISTED_PLATFORMS.test(newDomain) && !isSimilarDomain(this.domain, newDomain)) {
            console.warn("link discarded, diverging domain names:", this.domain, newDomain);
            return;
        }

        if (DETECT_SITEMAP_LINK.test(link)) {
            this.sitemapUrls.push(link);
        } else {
            this.urls.push(link);
        }
    }

    extractLinks(regex, index, handler) {
        let match;
        let count = 0;
        while ((match = regex.exec(this.content)) !== null && count < MAX_LINKS) {
            handler(match[index]);
            count++;
        }
        console.debug(
            `${this.sitemapUrls.length} sitemaps and ${this.urls.length} links found for ${this.currentUrl}`
        );
    }

    extractSitemapLanglinks() {
        if (!this.content.includes("hreflang=")) return;

        var langRegex = new RegExp(`hreflang=['"](?:${this.targetLang}.*?|x-default)['"]`, 'g');

        var handleLangLink = (attrs) => {
            if (langRegex.test(attrs)) {
                var langMatch = HREFLANG_REGEX.exec(attrs);
                if (langMatch) {
                    this.handleLink(langMatch[1]);
                }
            }
        };

        this.extractLinks(XHTML_REGEX, 0, handleLangLink);
    }

    extractSitemapLinks() {
        this.extractLinks(LINK_REGEX, 1, this.handleLink.bind(this));
    }

    process() {
        var plausible = isPlausibleSitemap(this.currentUrl, this.content);
        if (!plausible) return;

        if (!SITEMAP_FORMAT.test(this.content)) {
            this.extractLinks(DETECT_LINKS, 0, this.handleLink.bind(this));
            return;
        }

        if (this.targetLang !== null) {
            this.extractSitemapLanglinks();
            if (this.sitemapUrls.length > 0 || this.urls.length > 0) return;
        }
        this.extractSitemapLinks();
    }
}

async function sitemapSearch(url, targetLang = null, external = false, sleepTime = 2000) {
    var [domainname, baseurl] = getHostinfo(url);
    if (!domainname) {
        console.warn("invalid URL:", url);
        return [];
    }

    if (!await isLivePage(baseurl)) {
        console.warn("base URL unreachable, dropping sitemap:", url);
        return [];
    }

    let urlfilter = null;

    let sitemapurls = url.endsWith('.gz') || url.endsWith('sitemap') || url.endsWith('.xml') ? [url] : [];

    if (url.length > baseurl.length + 2) {
        urlfilter = url;
    }

    var sitemap = new SitemapObject(baseurl, domainname, sitemapurls, targetLang, external);

    if (sitemap.sitemapUrls.length === 0) {
        sitemap.sitemapUrls = await findRobotsSitemaps(baseurl) || GUESSES.map(g => `${baseurl}/${g}`);
    }

    while (sitemap.sitemapUrls.length > 0 && sitemap.seen.size < MAX_SITEMAPS_SEEN) {
        sitemap.currentUrl = sitemap.sitemapUrls.pop();
        await sitemap.fetch();
        sitemap.process();
        sitemap.sitemapUrls = sitemap.sitemapUrls.filter(s => !sitemap.seen.has(s));

        if (sitemap.seen.size < MAX_SITEMAPS_SEEN) {
            await new Promise(resolve => setTimeout(resolve, sleepTime));
        }
    }

    if (urlfilter) {
        sitemap.urls = filterUrls(sitemap.urls, urlfilter);
    }

    console.debug(`${sitemap.urls.length} sitemap links found for ${domainname}`);
    return sitemap.urls;
}

function isPlausibleSitemap(url, contents) {
    if (!contents) return false;

    url = url.replace(SCRUB_REGEX, '');

    if ((POTENTIAL_SITEMAP.test(url) && !SITEMAP_FORMAT.test(contents)) || contents.toLowerCase().includes('<html')) {
        console.warn("not a valid XML sitemap:", url);
        return false;
    }

    return true;
}

async function findRobotsSitemaps(baseurl) {
    var robotstxt = await fetchUrl(baseurl + "/robots.txt");
    return extractRobotsSitemaps(robotstxt, baseurl);
}

function extractRobotsSitemaps(robotstxt, baseurl) {
    if (!robotstxt || robotstxt.length > 10000) return [];

    var sitemapurls = [];
    var lines = robotstxt.split('\n');

    for (var line of lines) {
        var trimmedLine = line.split('#')[0].trim();
        if (!trimmedLine) continue;

        var [key, value] = trimmedLine.split(':', 2).map(s => s.trim());
        if (key && value && key.toLowerCase() === 'sitemap') {
            var candidate = fixRelativeUrls(baseurl, value);
            sitemapurls.push(candidate);
        }
    }

    console.debug(`${sitemapurls.length} sitemaps found in robots.txt`);
    return sitemapurls;
}

module.exports = {
    sitemapSearch,
    isPlausibleSitemap,
    findRobotsSitemaps,
    extractRobotsSitemaps
};